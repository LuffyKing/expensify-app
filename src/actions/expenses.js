import uuid from 'uuid';
import database from '../firebase/firebase'
//ADD expense
export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
}

};

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) =>{
    const uid = getState().auth.uid;
    const {
      description='',
      note='',
      amount=0,
      createdAt=0,
    } = expenseData;
    const expense = {description, note, amount, createdAt};
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id:ref.key,
        ...expense
      }));
    });
  };
;}
//  REMOVE_EXPENSE
export const removeExpense = ({id}) => {
  return {
    type:'REMOVE_EXPENSE',
    id

  }
};
//  EDIT_EXPENSE
export const editExpense = (id, updates) => {
  return {
    type:'EDIT_EXPENSE',
    id,
    updates,
  }
};
// set_expenses
export const setExpenses = (expenses) => {
  return {  type:'SET_EXPENSES',
  expenses}
};
//export const startSetExpenses
export const startSetExpenses = () => {
  return (dispatch, getState) =>{
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot)=>{
        expenses.push({
          id:childSnapshot.key,
          ...childSnapshot.val()
        });
      });
    dispatch(setExpenses(expenses));
  }).catch((e)=>{
    console.log('The following error occured',e)
  });
}
}
export const startRemoveExpense = ({id}) => {
  return (dispatch, getState) =>{
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {

      dispatch(removeExpense({id}));
    }).catch((e)=>{
      console.log('The following error occured',e)
    });
  }
};
export const startEditExpense = (id, updates) =>{
  return (dispatch, getState) =>{
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses/${id}`).update({
        ...updates
      }).then(() => {
      dispatch(editExpense(id, updates));
    }).catch((e)=>{
      console.log('The following error occured',e);
    });
  };
};
