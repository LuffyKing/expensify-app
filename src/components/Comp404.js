import React from 'react';
import {Link} from 'react-router-dom';
const Comp404 = (props) => {
  return (
    <div>404!
      <Link to="/">Go Home</Link>
    </div>
  )
}

export default Comp404;
