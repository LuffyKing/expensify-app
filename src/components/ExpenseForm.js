import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';
const now = moment();
class ExpenseForm extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      description: props.expense ? props.expense.description:'',
      note: props.expense ? props.expense.note:'',
      amount: props.expense ? props.expense.amount:'',
      createdAt: props.expense ? moment(props.expense.createdAt):moment(),
      calendarFocused:false,
      error:''

    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));

  };
  onTextAreaChange = (e) => {
    const note = e.target.value;
    this.setState(()=>({note}));

  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    const numTest = new RegExp(/^[1-9]\d*\.?\d{0,2}$/, 'i');
    if(numTest.test(amount)||!amount){
      this.setState(() => ({amount}));
    };


  };
  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(()=>({createdAt}));
    }
  };
  onFocusChange= ({focused}) =>{
    this.setState(() => ({calendarFocused: focused}));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if(!this.state.description||!this.state.amount){
      const error='Please provide description and amount!';
      this.setState(() =>({error}));
    }
    else{
      const error='';
      this.setState(() =>({error}));
      this.props.onSubmit({
        description: this.state.description,
        amount: +this.state.amount,
        note:this.state.note,
        createdAt:this.state.createdAt.valueOf(),
        note:this.state.note,
      })
    }

  };
  render(){
    return(

        <form onSubmit={this.onSubmit} className='form'>
          {

            this.state.error&& <h3>Error!! {this.state.error}</h3>
          }
          <input
            type="text"
            autoFocus
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            className="text-input"
          />

          <input
            type="text"
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder="Amount"/>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            >
          </SingleDatePicker>
          <textarea
            placeholder="Add a note for your expense (optional)"
            value = {this.state.note}
            onChange={this.onTextAreaChange}
            className="textarea"
            >

          </textarea>
          <div><button className="genButton">Save Expense</button></div>
        </form>
    );
  };
}
export default ExpenseForm;
