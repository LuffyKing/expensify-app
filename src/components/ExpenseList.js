import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from '../components/ExpenseListItem';
import selectExpenses from '../selector/expenses';
export const ExpenseList = (props) =>{
  return(
    <div>
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {
          props.expenses.length === 0 ? (
            <div className="list-item list-item--message">No Expenses</div>
          ): props.expenses.map((expense, index)=>{
            index+=1;

            return(

              <ExpenseListItem {...expense} index={index} key={expense.id}></ExpenseListItem>
            )
          })
        }
      </div>

    </div>
  );
};

const mapStateToProps  = (state) => {
  return {
     expenses: selectExpenses(state.expenses, state.filters)
  };
};


export default connect(mapStateToProps)(ExpenseList);
