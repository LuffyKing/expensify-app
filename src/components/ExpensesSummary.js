import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selector/expenses';
import selectExpensesTotal from '../selector/expenses-total';
import numeral from 'numeral';

export class ExpenseSummary extends React.Component{
  expenseCountF = () => {
    const expText = this.props.expenseCount === 1 ? 'expense':'expenses';
    return `${this.props.expenseCount} ${expText}`;
  }
  expenseAmountF = () => {
    const amoResult = this.props.expenseTotal;
    return `${numeral(amoResult).format('$0,0.00')}`;
  }
  render(){
    return(
      <p>
        Viewing {this.expenseCountF()}  totalling {this.expenseAmountF()}
      </p>

    );
  }


}


const mapStateToProps  = (state) => {
  return {
     expenses: selectExpenses(state.expenses, state.filters),
     expenseCount:state.expenses.length,
     expenseTotal:selectExpensesTotal(state.expenses)

  };
};

// const ExpenseSummary = (props) =>{
//   return(
//     <p>Viewing {props.expenses.length} {props.expenses.length===1?'expense':'expenses'}  totalling {selectExpensesTotal(props.expenses)}</p>
//   )
// }
// const mapStateToProps  = (state) => {
//   return {
//      expenses: selectExpenses(state.expenses, state.filters)
//   };
// };

export default connect(mapStateToProps)(ExpenseSummary);
