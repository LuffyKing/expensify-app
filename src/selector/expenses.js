import moment from 'moment';
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day'): true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day'):true;
    const regTest = new RegExp(text, 'i');

    const textMatch = ~expense.description.search(regTest);
    return startDateMatch && endDateMatch && textMatch;


  }).sort((expense1, expense2)=>{
    if(sortBy==='date'){
      return -1*(expense1.createdAt - expense2.createdAt);
    }else if (sortBy==='amount') {
      return -1*(expense1.amount - expense2.amount);

    }


  });

};
export default getVisibleExpenses;
