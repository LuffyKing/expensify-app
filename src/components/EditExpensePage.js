import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{
  onRemove= () => {

    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  render(){
    return(
      <div>

        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}></ExpenseForm>
        <button
          onClick={this.onRemove}
          >
        Remove
      </button>

      </div>

    );
  }


}

// const EditExpensePage = (props) => {
//   return (
    // <div>
    //   Editing expense with id of {props.match.params.id}
    //   <ExpenseForm expense={props.expense} onSubmit={(expense)=>{
    //     props.dispatch(editExpense(props.match.params.id, expense));
    //     props.history.push('/');
    //   }}></ExpenseForm>
    //   <button
    //     onClick={() => {
    //       console.log(props);
    //       let id = props.match.params.id;
    //       props.dispatch(removeExpense({id}));
    //       props.history.push('/');
    //    }
    // }>
    //   Remove
    // </button>
    //
    // </div>
//   )
// }

const mapDispatchToProps = (dispatch)=>{
  return {
    removeExpense:(id)=>{return dispatch(removeExpense({id}))},
    editExpense:(id, expense)=>dispatch(editExpense(id, expense))
  };
};

const mapStateToProps = (state, props) => {
  return{
    expense:state.expenses.find((expense) => {
          return expense.id === props.match.params.id;
    })
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
