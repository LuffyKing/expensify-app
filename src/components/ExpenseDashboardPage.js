import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpensesSummary';
const ExpenseDashboardPage = () => (
  <div>
    <ExpenseSummary></ExpenseSummary>
    <ExpenseListFilters></ExpenseListFilters>
    <ExpenseList></ExpenseList>
  </div>
);
export default ExpenseDashboardPage;
