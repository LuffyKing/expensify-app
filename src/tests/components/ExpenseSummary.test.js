import React from 'react';
import {shallow} from 'enzyme';
import  {ExpenseSummary} from '../../components/ExpensesSummary';
import  selectExpensesTotal from '../../selector/expenses-total';
import expenses from '../fixtures/expenses';
import numeral from 'numeral';
let wrapper;
beforeEach(() => {
  wrapper = shallow(<ExpenseSummary expenseCount={0} expenseTotal={0}></ExpenseSummary>);

});
test('this test if the component renders', ()=>{
  expect(wrapper).toMatchSnapshot();
});
test('this test for one expense', ()=>{
  wrapper.setProps({
    expenseCount:[expenses[0]].length,
    expenseTotal:selectExpensesTotal([expenses[0]])
  });
  expect(wrapper).toMatchSnapshot();
});
test('this test for three expenses', ()=>{
  wrapper.setProps({
    expenseCount:expenses.length,
    expenseTotal:selectExpensesTotal(expenses)
  });
  expect(wrapper).toMatchSnapshot();
});
