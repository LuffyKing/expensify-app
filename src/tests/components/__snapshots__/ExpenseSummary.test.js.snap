// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`this test if the component renders 1`] = `
<p>
  Viewing 
  undefined expenses
    totalling 
  $0.00
</p>
`;
